workflow/src/main/java/bio/terra/tanagra/workflow/PrecomputeCounts.java

```diff
@@ -0,0 +1,195 @@
+ *
+ * <p>primary nodes (node) = list of (condition)'s
+ *
+ * <p>occurrences (node,what_to_count) = list of condition_occurrence rows (condition,person)
+ *
+ * <p>ancestor descendant relationships (ancestor,descendant) = list of ancestor-descendant pairs in
+ * the condition hierarchy, as generated by the ({@link FlattenHierarchy}) workflow
+ *
+ * <p>OUTPUT: counts (node,count) = (condition, # people with >=1 occurrence of the condition or any
+ * of its children)
+ */
+public final class PrecomputeCounts {
+  private PrecomputeCounts() {}
+
+  /** Options supported by {@link PrecomputeCounts}. */
+  public interface PrecomputeCountsOptions extends BigQueryOptions {
+    @Description(
+        "Path to a BigQuery standard SQL query file to execute to retrieve all primary entity instances. "
+            + "The result of the query should have one column, (node).")
+    String getAllPrimaryNodesQuery();
+
+    void setAllPrimaryNodesQuery(String query);
+
+    String getAllPrimaryNodesQueryText();
+
+    void setAllPrimaryNodesQueryText(String query);
+
+    @Description(
+        "Path to a BigQuery standard SQL query file to execute to retrieve all occurrence entity instances. "
+            + "The result of the query should have two columns, (node, what_to_count).")
+    String getOccurrencesQuery();
+
+    void setOccurrencesQuery(String query);
+
+    String getOccurrencesQueryText();
+
+    void setOccurrencesQueryText(String query);
+
+    @Description(
+        "Path to a BigQuery standard SQL query file to execute to retrieve all ancestor-descendant relationships for the primary entity. "
+            + "The result of the query should have two columns, (ancestor, descendant).")
+    String getAncestorDescendantRelationshipsQuery();
+
+    void setAncestorDescendantRelationshipsQuery(String query);
+
+    String getAncestorDescendantRelationshipsQueryText();
+
+    void setAncestorDescendantRelationshipsQueryText(String query);
+
+    @Description(
+        "The  \"[project_id]:[dataset_id].[table_id]\" specification of the precomputed-counts table to create.")
+    String getOutputBigQueryTable();
+
+    void setOutputBigQueryTable(String tableRef);
+
+    @Description("What to name the node column on the output BigQuery table.")
+    @Default.String("node")
+    String getOutputNodeColumn();
+
+    void setOutputNodeColumn(String nodeColumn);
+
+    @Description("What to name the count column on the output BigQuery table.")
+    @Default.String("count")
+    String getOutputCountColumn();
+
+    void setOutputCountColumn(String countColumn);
+  }
+
+  public static void main(String[] args) throws IOException {
+    PrecomputeCountsOptions options =
+        PipelineOptionsFactory.fromArgs(args).withValidation().as(PrecomputeCountsOptions.class);
+    Pipeline pipeline = Pipeline.create(options);
+
+    // read in the queries from files
+    String primaryNodesQuery = Files.readString(Path.of(options.getAllPrimaryNodesQuery()));
```

Comment:
You could consider structuring this like I did for WriteAllNodes. That moves the queries and other configuration into the underlay rather than separate files so it's easy to run using RunUnderlayWorkflows, even if there's not actually any smarter logic there yet.

Reply:
Yes, I'm planning to refactor all the Beam workflows together in a separate PR.